### PR TITLE
feat(lua)!: execute Lua with "nvim -l"

### LINKED ISSUE #15749: startup: "-l" enables Lua mode
`nvim -l` should enable "Lua mode" during startup, i.e.:

1. ✅ stdin can be executed as Lua?
2. ✅ args are copied to the Lua magic `_G.arg` global, for use in scripts
    - "negative args" contain all other `nvim` args?
3. ~~`-u {file}`is executed as the main Lua script _regardless of file extension_.~~

I see at least 2 ways of doing this:

* ~~`-l` strictly enables "Lua mode", and `-u foo.lua` specifies the file.~~
* ✅ `-l foo.lua` works, and also enables "Lua mode".
* ✅ All other `nvim` args can be used also (until `-l`). So you can mix vimscript `-c`/`+` with `-l`, load files into buffers, etc.

"Lua mode" could work like this (other suggestions welcome):
* ~~all args after `--` are sent to the Lua magic `arg` global.~~
* ✅  all args after `-l` are sent to the Lua magic `arg` global. 

Other:

- Tradtionally/conventionally, the [`lua` CLI](https://www.lua.org/pil/1.4.html) sets _negative_ args to CLI args _not_ sent to the "main" script. 
    - But this is awkward to work with in Lua. We already have `v:argv`, so scripts can inspect that if they care about the full arg list.
    - We will break `lua` CLI tradition here unless a use-case is presented.
- > one more thing: can the script control if the "normal" editor loop is started or not after its done? sometimes you want to run a one-shot script using nvim's builtin libraries and then exit, some other times the script would set up the editor to do something and then hand off to the main loop after the script is node.
    - ❌  Not planned. Workarounds:
        - Use `-c "luafile ..."`
        - Use `-S foo.lua -S bar.lua` to run multiple Lua scripts.

### PR DESCRIPTION
**Pull request by justinmk**
# Examples

```
nvim  bar.txt -l foo.lua
echo 'print(vim.inspect(vim.api.nvim_buf_get_text(1,0,0,-1,-1,{})))' | nvim +"put ='text'" -l -
echo "print(vim.inspect(_G.arg))" | nvim -l - --arg1 --arg2
```

closes #15749

# Todo

- [x] tests for shada/swap/userconfig
- [x] exit 1 on lua error
- [x] exit without `+q`
- [x] update docs-gen CI job
- [x] support `-l -` (stdin)
- [x] fix: os_msg, os_errmsg should not truncate based on screen width! #24216
- [ ] need some sort of `vim.is_main()` magic function that detects whether the current module was executed by `nvim -l`.
    - somewhat mitigated by `_G.arg[0]` #21663
- [ ] adjust places that test `verbose==1`
- [ ] ensure final `\n` in output?

# Future

To run tests with `nvim` instead of `lua` or `luajit`, change `.deps/usr/bin/busted` script as follows:

    #!/bin/sh

    LUAROCKS_SYSCONFDIR='/…/.deps/usr/etc/luarocks' exec '/…/build/bin/nvim' -c 'lua package.path="/…/.deps/usr/share/lua/5.1/?.lua;/…/.deps/usr/share/lua/5.1/?/init.lua;"..package.path;package.cpath="/…/.deps/usr/lib/lua/5.1/?.so;"..package.cpath;local k,l,_=pcall(require,"luarocks.loader") _=k and l.add_context("busted","2.0.0-1")' -l '/…/.deps/usr/lib/luarocks/rocks-5.1/busted/2.0.0-1/bin/busted' "$@"

With this PR, plus this patch:

```diff
diff --git a/test/functional/helpers.lua b/test/functional/helpers.lua
index ca59eb31828b..a01d62af98e9 100644
--- a/test/functional/helpers.lua
+++ b/test/functional/helpers.lua
@@ -957,7 +957,7 @@ function module.mkdir_p(path)
     or 'mkdir -p '..path))
 end
 
-module = global_helpers.tbl_extend('error', module, global_helpers)
+module = global_helpers.tbl_extend('keep', module, global_helpers)
 
 return function(after_each)
   if after_each then
diff --git a/test/helpers.lua b/test/helpers.lua
index 3fe4322501a3..320b826b6da5 100644
--- a/test/helpers.lua
+++ b/test/helpers.lua
@@ -811,6 +811,6 @@ function module.read_nvim_log(logfile, ci_rename)
   return log
 end
 
-module = shared.tbl_extend('error', module, Paths, shared, require('test.deprecated'))
+module = shared.tbl_extend('keep', module, Paths, shared, require('test.deprecated'))
 
 return module
```

tests run successfully using `nvim` as the test runner:

```
|| -------- Running tests from test/functional/core/startup_spec.lua
|| RUN      T1 startup -l Lua failure modes: 23.30 ms OK
|| RUN      T2 startup -l Lua os.exit() sets Nvim exitcode: 39.30 ms OK
|| RUN      T3 startup -l Lua sets _G.arg: 189.28 ms OK
|| -------- 3 tests from test/functional/core/startup_spec.lua (262.84 ms total)
|| 
|| -------- Global test environment teardown.
|| ======== 3 tests from 1 test file ran. (263.25 ms total)
|| PASSED   3 tests.
```

_and_ fail correctly:

```
|| ERROR    test/functional/core/startup_spec.lua @ 99: startup -l Lua failure modes
|| test/helpers.lua:85: Pattern does not match.
|| Pattern:
|| nvim: XArgument missing after: "%-l"
|| Actual:
|| nvim: Argument missing after: "-l"
|| More info with "nvim -h"
```

### CODE REVIEW

**Comment by justinmk** [comment]
test failing on windows... is there a problem with getting output from `system(['nvim', '-es', ...])` there?

**Comment by justinmk** [comment]
going to merge this (followup PRs later) once I can fix the Windows CI. (something to do with capturing stderr?)

**Comment by zeertzjq** [comment]
Maybe some Vim patches are missing for `os_errmsg()` on Windows?
- https://github.com/vim/vim/commit/9b5c1fcdeae75f82a2083fafbbf75ab220f6ac1e
- https://github.com/vim/vim/commit/97c2c05ead689f4a5de986c82531fcbd09295cda
- https://github.com/vim/vim/commit/a97c36310f90ed15dbf5a2ba5bf91fc906e2e724

**Comment by justinmk** [comment]
> Maybe some Vim patches are missing for `os_msg()` and `os_errmsg()` on Windows?

thank you for the help! actually Nvim was crashing (bug in `nlua_set_argv`) in the nested system() call, which returned empty output. probably can make this better by checking `v:shell_error` in `assert_l_out` before checking the output.

**Comment by justinmk** [comment]
after merging this,

- `regen-api-docs` job will fail until nightly has this feature
- we can use `nvim` as the test-runner instead of `lua` 
    - eliminates most of the test util functions in `helpers.lua`.
    - see the "Future" note above.

File: src/nvim/message.c [hunk 1]
```diff
@@ -303,12 +303,14 @@ bool msg_attr_keep(const char *s, int attr, bool keep, bool multiline)
     // the correct time already).
     abort();
   }
+  ELOG("www 1: here");
```
[the remark this answers is not in the document]
Comment (justinmk) [reply src/nvim/message.c:306]: yeah, still cleanup to do

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -281,6 +283,8 @@ jobs:
       DEPS_BUILD_DIR: ${{ github.workspace }}/nvim-deps
       CACHE_NVIM_DEPS_DIR: ${{ github.workspace }}/nvim-deps
       DEPS_PREFIX: ${{ github.workspace }}/nvim-deps/usr
+      # TEST_FILE: test/functional/core/startup_spec.lua
+      # TEST_FILTER: foo
```
Comment (justinmk) [inline .github/workflows/ci.yml:287]: Leaving this here and documented in CONTRIBUTING.md to help others.

**Comment by justinmk** [comment]
ci failure looks unrelated:

```
FAILED   test/functional/terminal/tui_spec.lua @ 2372: TUI as a client throws error when no server exists
test/functional/terminal/tui_spec.lua:2380: Row 2 did not match.
Expected:
  |Remote ui failed to start: {MATCH:.*}|
  |*                                                            |
  |*[Process exited 1]{1: }                                         |
  |*                                                            |
  |                                                            |
  |                                                            |
  |{3:-- TERMINAL --}                                              |
Actual:
  |Remote ui failed to start: failed to lookup host o          |
  |*r port                                                      |
  |*                                                            |
  |*[Process exited 1]{1: }                                         |
  |                                                            |
  |                                                            |
  |{3:-- TERMINAL --}                                              |
```

File: .github/workflows/ci.yml (hunk 2 above)
Comment (dundargoc) [reply .github/workflows/ci.yml:287]: Just put one set of environment variables at the top level instead?

Comment (justinmk) [reply .github/workflows/ci.yml:287]: That didn't seem to work when I tried it...

**Pull request merged**